RESTEasy Reactive build failure with MULTIPART_FORM_DATA

### Describe the bug

Using `quarkus-resteasy-reactive` extension, I'm trying to do this:
```
@Path("/echo")
public class MyResource {
    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Produces(MediaType.TEXT_PLAIN)
    public String echo(byte[] request) {
        return new String(request, StandardCharsets.UTF_8);
    }
}
```
which should be fine (works for non-reactive `quarkus-resteasy`).
But when I invoke `mvn package`, I get an exception:
```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.1.2.Final:build (default) on project my-artifactId: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor#setupEndpoints threw an exception: java.lang.RuntimeException: java.lang.RuntimeException: Failed to process method org.my.group.MyResource#java.lang.String echo(byte[] request)
[ERROR] 	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:245)
[ERROR] 	at io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor.setupEndpoints(ResteasyReactiveProcessor.java:423)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.lang.RuntimeException: Failed to process method org.my.group.MyResource#java.lang.String echo(byte[] request)
[ERROR] 	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod(EndpointIndexer.java:548)
[ERROR] 	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:293)
[ERROR] 	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:216)
[ERROR] 	... 12 more
[ERROR] Caused by: java.lang.NullPointerException
[ERROR] 	at io.quarkus.resteasy.reactive.server.deployment.QuarkusServerEndpointIndexer.handleMultipart(QuarkusServerEndpointIndexer.java:226)
[ERROR] 	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod(EndpointIndexer.java:459)
[ERROR] 	... 14 more
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

[resteasy-reactive-multipart-reproducer.zip](https://github.com/quarkusio/quarkus/files/6988004/resteasy-reactive-multipart-reproducer.zip)
```
unzip ./resteasy-reactive-multipart-reproducer.zip
cd ./resteasy-reactive-multipart-reproducer/
./mvnw clean package
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "11.0.11" 2021-04-20 OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9) OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.1.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)

### Additional information

_No response_

--- a/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
+++ b/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
@@ -958,7 +958,14 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
                     // in this case it is safe to assume that we are consuming multipart data
                     // we already don't allow multipart to be used along with body in the same method,
                     // so this is completely safe
-                    builder.setType(ParameterType.MULTI_PART_FORM);
+                    var type = toClassName(paramType, currentClassInfo, actualEndpointInfo, index);
+                    var typeInfo = index.getClassByName(DotName.createSimple(type));
+                    if (typeInfo != null && typeInfo.annotations().containsKey(REST_FORM_PARAM)) {
+                        builder.setType(ParameterType.MULTI_PART_FORM);
+                    } else {
+                        //if the paramater does not have @RestForm annotations we treat it as a normal body
+                        builder.setType(ParameterType.BODY);
+                    }
                 } else {
                     builder.setType(ParameterType.BODY);
                 }

--- a/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/model/ResourceReader.java
+++ b/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/model/ResourceReader.java
@@ -21,8 +21,9 @@ public class ResourceReader {
     private volatile List<MediaType> mediaTypes;
     private volatile MessageBodyReader<?> instance;
 
-    public void setFactory(BeanFactory<MessageBodyReader<?>> factory) {
+    public ResourceReader setFactory(BeanFactory<MessageBodyReader<?>> factory) {
         this.factory = factory;
+        return this;
     }
 
     public BeanFactory<MessageBodyReader<?>> getFactory() {

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/framework/ResteasyReactiveUnitTest.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/framework/ResteasyReactiveUnitTest.java
@@ -31,12 +31,14 @@ import java.util.logging.LogRecord;
 import java.util.logging.Logger;
 import javax.ws.rs.core.Application;
 import javax.ws.rs.core.MediaType;
+import javax.ws.rs.ext.MessageBodyReader;
 import javax.ws.rs.ext.MessageBodyWriter;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.Index;
 import org.jboss.resteasy.reactive.common.ResteasyReactiveConfig;
 import org.jboss.resteasy.reactive.common.model.ResourceClass;
+import org.jboss.resteasy.reactive.common.model.ResourceReader;
 import org.jboss.resteasy.reactive.common.model.ResourceWriter;
 import org.jboss.resteasy.reactive.common.processor.JandexUtil;
 import org.jboss.resteasy.reactive.common.processor.scanning.ApplicationScanningResult;
@@ -46,6 +48,7 @@ import org.jboss.resteasy.reactive.common.processor.scanning.ResteasyReactiveSca
 import org.jboss.resteasy.reactive.server.core.Deployment;
 import org.jboss.resteasy.reactive.server.core.DeploymentInfo;
 import org.jboss.resteasy.reactive.server.core.ServerSerialisers;
+import org.jboss.resteasy.reactive.server.core.startup.CustomServerRestHandlers;
 import org.jboss.resteasy.reactive.server.core.startup.RuntimeDeploymentManager;
 import org.jboss.resteasy.reactive.server.handlers.RestInitialHandler;
 import org.jboss.resteasy.reactive.server.processor.ServerEndpointIndexer;
@@ -53,6 +56,7 @@ import org.jboss.resteasy.reactive.server.processor.scanning.ResteasyReactiveCon
 import org.jboss.resteasy.reactive.server.processor.scanning.ResteasyReactiveExceptionMappingScanner;
 import org.jboss.resteasy.reactive.server.processor.scanning.ResteasyReactiveFeatureScanner;
 import org.jboss.resteasy.reactive.server.processor.scanning.ResteasyReactiveParamConverterScanner;
+import org.jboss.resteasy.reactive.server.providers.serialisers.ServerByteArrayMessageBodyHandler;
 import org.jboss.resteasy.reactive.server.providers.serialisers.ServerStringMessageBodyHandler;
 import org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler;
 import org.jboss.resteasy.reactive.server.vertx.VertxRequestContextFactory;
@@ -252,8 +256,27 @@ public class ResteasyReactiveUnitTest implements BeforeAllCallback, AfterAllCall
                         };
                     }
                 }));
+        serialisers.addReader(byte[].class, new ResourceReader()
+                .setMediaTypeStrings(Collections.singletonList(MediaType.WILDCARD))
+                .setFactory(new BeanFactory<MessageBodyReader<?>>() {
+                    @Override
+                    public BeanInstance<MessageBodyReader<?>> createInstance() {
+                        return new BeanInstance<MessageBodyReader<?>>() {
+                            @Override
+                            public MessageBodyReader<?> getInstance() {
+                                return new ServerByteArrayMessageBodyHandler();
+                            }
+
+                            @Override
+                            public void close() {
+
+                            }
+                        };
+                    }
+                }));
         DeploymentInfo info = new DeploymentInfo()
                 .setApplicationPath("/")
+                .setConfig(new ResteasyReactiveConfig())
                 .setFeatures(ResteasyReactiveFeatureScanner.createFeatures(index, applicationScanningResult))
                 .setInterceptors(
                         ResteasyReactiveInterceptorScanner.createResourceInterceptors(index, applicationScanningResult))
@@ -284,7 +307,8 @@ public class ResteasyReactiveUnitTest implements BeforeAllCallback, AfterAllCall
                         }
                     }
                 });
-        RuntimeDeploymentManager runtimeDeploymentManager = new RuntimeDeploymentManager(info, () -> executor, null,
+        RuntimeDeploymentManager runtimeDeploymentManager = new RuntimeDeploymentManager(info, () -> executor,
+                new CustomServerRestHandlers(null),
                 closeable -> closeTasks.add(closeable), new VertxRequestContextFactory(), ThreadSetupAction.NOOP, "/");
         Deployment deployment = runtimeDeploymentManager.deploy();
         RestInitialHandler initialHandler = new RestInitialHandler(deployment);

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/multipart/EchoResource.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/multipart/EchoResource.java
@@ -1,0 +1,21 @@
+package org.jboss.resteasy.reactive.server.vertx.test.multipart;
+
+import io.smallrye.common.annotation.NonBlocking;
+import java.nio.charset.StandardCharsets;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/echo")
+public class EchoResource {
+
+    @POST
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    @Produces(MediaType.TEXT_PLAIN)
+    @NonBlocking
+    public String echo(byte[] request) {
+        return new String(request, StandardCharsets.UTF_8);
+    }
+}

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/multipart/RawMultipartTestCase.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/multipart/RawMultipartTestCase.java
@@ -1,0 +1,34 @@
+package org.jboss.resteasy.reactive.server.vertx.test.multipart;
+
+import io.restassured.RestAssured;
+import java.util.function.Supplier;
+import org.hamcrest.Matchers;
+import org.jboss.resteasy.reactive.server.vertx.test.framework.ResteasyReactiveUnitTest;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class RawMultipartTestCase {
+
+    @RegisterExtension
+    static ResteasyReactiveUnitTest test = new ResteasyReactiveUnitTest()
+            .setArchiveProducer(new Supplier<JavaArchive>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class).addClass(EchoResource.class);
+                }
+            });
+
+    @Test
+    public void testHelloEndpoint() {
+        RestAssured.given()
+                .when()
+                .multiPart("testName", "testValue")
+                .post("/echo")
+                .then()
+                .statusCode(200)
+                .body(Matchers.containsString("testValue"));
+    }
+
+}

